src/librustc_target/abi/call/riscv.rs

```diff
@@ -1,49 +1,341 @@
 // Reference: RISC-V ELF psABI specification
 // https://github.com/riscv/riscv-elf-psabi-doc
+//
+// Reference: Clang RISC-V ELF psABI lowering code
+// https://github.com/llvm/llvm-project/blob/8e780252a7284be45cf1ba224cabd884847e8e92/clang/lib/CodeGen/TargetInfo.cpp#L9311-L9773
 
-use crate::abi::call::{ArgAbi, FnAbi};
+use crate::abi::call::{ArgAbi, ArgAttribute, FnAbi, PassMode, Reg, Uniform};
+use crate::abi::{self, Abi, FieldPlacement, LayoutOf, TyLayout, TyLayoutMethods};
+use crate::spec::HasTargetSpec;
+
+#[derive(Copy, Clone)]
+enum FieldKind {
+    Float,
+    Integer,
+    Unknown,
+}
+
+#[derive(Copy, Clone)]
+enum FloatArgPass {
+    FloatPair,
+    Float,
+    MixedPair,
+}
+
+fn is_vector<'a, Ty>(arg: &ArgAbi<'a, Ty>) -> bool {
+    match arg.layout.abi {
+        Abi::Vector { .. } => true,
+        _ => false,
+    }
+}
+
+fn is_aggregate<'a, Ty>(arg: &ArgAbi<'a, Ty>) -> bool {
+    arg.layout.is_aggregate() || is_vector(arg)
+}
+
+fn should_use_fp_conv_helper<'a, Ty, C>(
+    cx: &C,
+    arg_layout: &TyLayout<'a, Ty>,
+    xlen: u64,
+    flen: u64,
+    field1_kind: &mut FieldKind,
+    field2_kind: &mut FieldKind,
+) -> bool
+where
+    Ty: TyLayoutMethods<'a, C> + Copy,
+    C: LayoutOf<Ty = Ty, TyLayout = TyLayout<'a, Ty>>,
+{
+    match arg_layout.abi {
+        Abi::Scalar(ref scalar) => match scalar.value {
+            abi::Int(..) | abi::Pointer => {
+                if arg_layout.size.bits() > xlen {
+                    return false;
+                }
+                match (*field1_kind, *field2_kind) {
+                    (FieldKind::Unknown, _) => {
+                        *field1_kind = FieldKind::Integer;
+                        true
+                    }
+                    (FieldKind::Float, FieldKind::Unknown) => {
```

Comment:
So this computation relies on visiting fields in order?

What about large padding gaps, or reordered fields? The homogenous aggregate is conservative about those cases, but it *does* check for them.

Reply:
https://github.com/riscv/riscv-elf-psabi-doc/blob/master/riscv-elf.md#-hardware-floating-point-calling-convention
>  Attributes such as aligned or packed do not interfere with a struct's eligibility for being passed in registers according to the rules below

Reply:
What about reordered fields? The `x86_64` implementation only looks at offsets - you can iterate the fields by ascending offset, if you want, there's some code around that does that (since we need it for e.g. LLVM types).

Reply:
Where is such code exactly?